### ISSUE TITLE
 ephemeral_storage_local_ssd_config is a read only field for lssd machine types. When this field changes, which happens every apply, terraform will destroy the nodepool and recreate it

### ISSUE THREAD

**Issue by philip-harvey** [hashicorp/terraform-provider-google#17068]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

Terraform v1.6.6
provider registry.terraform.io/hashicorp/google v5.12.0

### Affected Resource(s)

google_container_node_pool

### Terraform Configuration Files

```tf
resource "google_container_node_pool" "nodepool" {
  provider = google-beta
....
dynamic "ephemeral_storage_local_ssd_config" { # new way to specify NVMe
      for_each = (
        var.node_local_nvme_ssd_count != null ? [var.node_local_nvme_ssd_count] : []
      )
      iterator = config
      content {
        local_ssd_count = config.value
      }
    }
...

```

### Debug Output
ephemeral_storage_local_ssd_config { # forces replacement
              - local_ssd_count = 1 -> null # forces replacement
            }

### Expected Behavior
This field is read only for lssd machine types, it can't be set and when it changes it should not trigger the nodepool to be deleted and recreated.
### Actual Behavior
Nodepool is deleted and recreated every terraform apply

### Steps to Reproduce
Use a lssd node type.  e.g. c3d-standard-8-lssd

1. `terraform apply`

### References
https://cloud.google.com/compute/docs/disks/local-ssd#choose_number_local_ssds

**Comment by edwardmedia** [comment]
encounter below error. Forwarding the issue to the api team. 
```
Error: error creating NodePool: googleapi: Error 400: The selected machine type (c3d-standard-8-lssd) has a fixed number of local SSD(s): 1. The EphemeralStorageConfig’s count field should be left unset or set to 1, but was set to 2.
```

**Comment by philip-harvey** [comment]
> encounter below error. Forwarding the issue to the api team.
> 
> ```
> Error: error creating NodePool: googleapi: Error 400: The selected machine type (c3d-standard-8-lssd) has a fixed number of local SSD(s): 1. The EphemeralStorageConfig’s count field should be left unset or set to 1, but was set to 2.
> ```

Hi Edward. This field is writable for many machine types, but read only for lssd machine types, which is somewhat troublesome for Terraform.   If the API just returned this as null would be expected, but instead it returns the number of NVMes that are assigned based on the machine type.

**Comment by dannawang0221** [comment]
Hi @philip-harvey, this is expected behavior that when you leave the local_ssd_count field unset, GKE backend will fill that field automatically with the number of local ssd disks according to the machine type. You can either set the local_ssd_count field explicitly with the number match to the machine type, for c3d-standard-8-lssd that would be 1, or you can remove the ephemeral_storage_local_ssd_config section, and GKE will configure local ssd disks as ephemeral storage by default for lssd machine types.

**Comment by philip-harvey** [comment]
> Hi @philip-harvey, this is expected behavior that when you leave the local_ssd_count field unset, GKE backend will fill that field automatically with the number of local ssd disks according to the machine type. You can either set the local_ssd_count field explicitly with the number match to the machine type, for c3d-standard-8-lssd that would be 1, or you can remove the ephemeral_storage_local_ssd_config section, and GKE will configure local ssd disks as ephemeral storage by default for lssd machine types.

Hi @dannawang0221   
I can't see any way for a module to handle a block that is sometimes writable and sometimes read only, and returns something when nothing is written to it.
The Terraform code 
dynamic "ephemeral_storage_local_ssd_config" { # new way to specify NVMe
      for_each = (
        var.node_local_nvme_ssd_count != null ? [var.node_local_nvme_ssd_count] : []
      )
      iterator = config
      content {
        local_ssd_count = config.value
      }
    }

would normally handle a dynamic block, but this is a really strange case and isn't handled by a dynamic block, which in turn causes the node pool to be deleted and recreated on every apply.  As a work around I've been specifying the number of SSDs that each lssd machine type will have attached, but this seems like a bug to have to specify the value for a read only field.

**Comment by dannawang0221** [comment]
Can you explain what does writable mean? If you use a non lssd machine type and update the local_ssd_count, I think the cluster/nodel_pool will also get destroy and recreate?

**Comment by philip-harvey** [comment]
> Can you explain what does writable mean? If you use a non lssd machine type and update the local_ssd_count, I think the cluster/nodel_pool will also get destroy and recreate?

True, but if you want to change the number of SSDs you can write to this field, hence it's writable.  For lssd machine types if you write to this field you get a 400 error.

**Comment by dannawang0221** [comment]
I see what you mean. The lssd machines like c3d-standard-8-lssd  are our 3rd generation machines, which have fixed number of lssd disks for each node, so you won't be able to set a different number, that's expected. For more details you can refer to https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/local-ssd#cluster

**Comment by philip-harvey** [comment]
> I see what you mean. The lssd machines like c3d-standard-8-lssd are our 3rd generation machines, which have fixed number of lssd disks for each node, so you won't be able to set a different number, that's expected. For more details you can refer to https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/local-ssd#cluster

Hi @dannawang0221   I understand that, but I can't find any way to make that work in the current version of the Google Terraform provider without having to use the hack of passing the number of SSDs, which shouldn't be required.

**Comment by marwanad** [comment]
@dannawang0221 any updates on this?

**Comment by Sneha-at** [comment]
Another possible alternative could be to ask Terraform to ignore any changes in the `local_ssd_count` field. Since you cannot update LSSDs after nodepool creation ([NodePool Update API](https://cloud.google.com/sdk/gcloud/reference/container/node-pools/update)) we don't have to monitor the LSSD count on any updates. 
```
resource "google_container_node_pool" "default_pool" {
  name       = "gke-node-pool"
  location   = var.region
  cluster     = google_container_cluster.gke_cluster.name
  node_config {
    machine_type = var.machine_type
    preemptible  = var.preemptible
    local_ssd_count = 0
  }

  lifecycle {
    ignore_changes = [
      # Ignore changes to local_ssd_count
      node_config.local_ssd_count,
    ]
  }
}
```

**Comment by jsok** [comment]
I'm hitting the same issue with my node pool being constantly re-created:

```
          - ephemeral_storage_local_ssd_config { # forces replacement
              - data_cache_count = 0 -> null
              - local_ssd_count  = 0 -> null
            }
```

My fix has been to use a nullable variable and dynamic block, since this module is used for node pools which don't have local SSDs:

```
variable "local_ssd_count" {
  type     = number
  default  = null
  nullable = true
}

resource "google_container_node_pool" "default" {

    dynamic "ephemeral_storage_local_ssd_config" {
      for_each = var.local_ssd_count == null ? [] : [true]
      content {
        data_cache_count = 0
        local_ssd_count  = var.local_ssd_count
      }
    }

}
```

**Comment by sunnylovestiramisu** [comment]
I wonder if we can add Computed: true to the schema for this field. 

In Terraform, Computed: true is a schema attribute that signifies that the value for a specific field is determined by the resource provider and cannot be known until after the resource has been created or updated. This is common for attributes that are automatically assigned by a cloud provider or external API. 

And when a user specifies a value for a field marked as Computed: true, the user's value always takes precedence.

For this case: The user's config is null, and the API returns {}.

Terraform compares the configuration (null) with the state ({}). Because the block is marked as Computed: true, Terraform's logic changes:

- The user did not specify this block in their configuration.
- However, the Computed flag tells me that the API is allowed to create it with its own default values.
- Therefore, the empty object {} in the state is an acceptable result.

The dual role of Optional: true + Computed: true is very common in Terraform providers for fields that have server-side defaults. Terraform's logic follows a clear priority.

We should experiment this option as a fix for the null case, but I am not sure about if it is set to 0.


But on the other hand why do you need to set local_ssd_count = 0? If the VM shape came with local SSD you cannot really override it with 0, this configuration should be removed entirely?

**Comment by carreter** [comment]
Following up on this. Experienced the delete & recreate issue when spinning up `a3-ultragpu-8g` instances, which come with local SSD by default.

I agree with @sunnylovestiramisu that this field should be marked as `Computed: true` in the schema, is there a strong reason for it not to be?
